trusted_os/flash.go

```diff
@@ -133,6 +146,61 @@ func determineLoadedOSBlock(card Card) error {
 	return nil
 }
 
+// newWitnessIdentity reads the newIdentityBlock from MMC.
+func newWitnessIdentity(card Card) (bool, error) {
+	blockSize := card.Info().BlockSize
+	if blockSize != expectedBlockSize {
+		return false, fmt.Errorf("h/w invariant error - expected MMC blocksize %d, found %d", expectedBlockSize, blockSize)
+	}
+
+	b, err := card.Read(newIdentityBlock*expectedBlockSize, 1)
+	if err != nil {
+		return false, err
+	}
+
+	buf := bytes.NewReader(b)
+	var newIdentity bool
+	if err := binary.Read(buf, binary.BigEndian, &newIdentity); err != nil {
+		return false, err
+	}
+
+	return newIdentity, nil
+}
+
+// incrementWitnessIdentityMMC increments the mmcIdentityCounter in the MMC block.
+func incrementWitnessIdentityMMC(card Card) error {
```

Comment:
What do you think about creating a `rpmb/rpmb_fake.go` file (with a `+fake_rpmb` build tag) which uses a few MMC blocks to emulate RPMB storage?

This func could move over to be the alternate implementation of the `RPMB.incrementWitnessIdentity` func which you're adding to the "real" rpmb implementation below (that `rpmb/rpmb.go` file would need a `!fake_rpmb` build tag).

We could add corresponding `witnessIDCounter()` funcs on the two implementations for returning the counter, too.

That way from the OS' PoV there would be no difference, and so we wouldn't need special cases to know which to call, nor two RPCs (as it currently is we would need the Applet to also somehow know which one to call).

Reply:
Thanks for the suggestion! It's much cleaner this way.